build.gradle

```diff
@@ -355,30 +356,38 @@ configure(javaProjects) {
         }
     }
 
-    task relocateSourceReferencesToInternalDependencies(type: ShadowJar) {
-        destinationDir = jar.destinationDir
-        archiveName = jar.baseName + '-relocated.jar'
+    task shadedJar(type: ShadowJar) {
+        destinationDir = tasks.jar.destinationDir
+        archiveName = "${tasks.jar.baseName}-shaded-${tasks.jar.version}.jar"
+
         // TODO(trustin): Uncomment this for faster building once shadow 2.0 is out.
         //                https://github.com/johnrengelman/shadow/issues/203
         //entryCompression = ZipEntryCompression.STORED
 
-        from sourceSets.main.output.classesDir
+        from sourceSets.main.output
         configurations = [ project.configurations.compile ]
 
         relocations.each {
-            relocate it[0], it[1]
+            relocate it[1], it[2]
         }
 
-        // Only want to rewrite source references, not bundle dependencies.
-        dependencies {
-            exclude(dependency({ true }))
+        exclude 'META-INF/maven/**'
+
+        if (project.name != 'core') {
+            // Only want to rewrite source references, not bundle dependencies.
+            dependencies {
+                exclude(dependency({ true }))
+            }
+        } else {
+            // Shade the relocated dependencies only.
+            dependencies {
+                exclude(dependency({ dep ->
+                    !relocations.find { dep.name.startsWith("${it[0]}:") }
+                }))
+            }
         }
     }
-    task extractRelocatedJar(type: Copy, dependsOn: relocateSourceReferencesToInternalDependencies) {
-        from zipTree(relocateSourceReferencesToInternalDependencies.archivePath)
-        into sourceSets.main.output.classesDir
-    }
-    jar.dependsOn extractRelocatedJar
+    jar.dependsOn tasks.shadedJar
```

Comment:
I think you can have build depend on shading instead of jar - it will make local tests faster.